src/goto-analyzer/static_simplifier.cpp

```diff
@@ -0,0 +1,303 @@
+
+// #define DEBUG
+
+#ifdef DEBUG
+#include <iostream>
+#endif
+
+#include <fstream>
+
+#include <analyses/interval_domain.h>
+#include <analyses/constant_propagator.h>
+
+#include <goto-programs/remove_skip.h>
+#include <goto-programs/remove_unreachable.h>
+#include <goto-programs/write_goto_binary.h>
+
+#include "static_simplifier.h"
+
+template<class analyzerT>
+class static_simplifiert:public messaget
+{
+public:
+  static_simplifiert(
+    goto_modelt &_goto_model,
+    const optionst &_options,
+    message_handlert &_message_handler,
+    std::ostream &_out):
+    messaget(_message_handler),
+    goto_functions(_goto_model.goto_functions),
+    ns(_goto_model.symbol_table),
+    options(_options),
+    out(_out)
+  {
+  }
+
+  bool operator()(void);
+
+protected:
+  goto_functionst &goto_functions;
+  const namespacet ns;
+  const optionst &options;
+  std::ostream &out;
+
+  // analyses
+  analyzerT domain;
+
+  void simplify_program(void);
+  unsigned simplify_guard(goto_programt::instructionst::iterator &i_it);
+};
+
+/*******************************************************************\
+
+Function: static_simplifiert<analyzerT>::operator()
+
+  Inputs: None.
+
+ Outputs: false on success, true on failure.
+
+ Purpose: Run the analysis, check the assertions and report in the
+          correct format.
+
+\*******************************************************************/
+
+template<class analyzerT>
+bool static_simplifiert<analyzerT>::operator()(void)
+{
+  status() << "Performing analysis" << eom;
+  domain(goto_functions, ns);
+
+  status() << "Simplifying program" << eom;
+  simplify_program();
+
+  // Remove obviously unreachable things and (now) unconditional branches
+  if(options.get_bool_option("simplify-slicing"))
+  {
+    status() << "Removing unreachable instructions" << eom;
+
+    remove_skip(goto_functions);  // Removes goto false
+    goto_functions.update();
+
+    remove_unreachable(goto_functions);  // Convert unreachable to skips
+    goto_functions.update();
+
+    remove_skip(goto_functions);  // Remove all of the new skips
+    goto_functions.update();
+  }
+
+  status() << "Writing goto binary" << eom;
+  return write_goto_binary(out, ns.get_symbol_table(), goto_functions);
+}
+
+
+/*******************************************************************\
+
+Function: static_simplifiert<analyzerT>::simplify_guard
+
+  Inputs: An iterator pointing to an instruction.
+
+ Outputs: 1 if simplified, 0 if not.
+
+ Purpose: Simplifies the instruction's guard using the information in
+          the abstract domain.
+
+\*******************************************************************/
+
+template<class analyzerT>
+unsigned static_simplifiert<analyzerT>::simplify_guard(
+  goto_programt::instructionst::iterator &i_it)
+{
+  exprt simplified=domain[i_it].domain_simplify(i_it->guard, ns);
+  unsigned return_value=(simplified==i_it->guard)?0:1;
+  i_it->guard=simplified;
+  return return_value;
```

Comment:
Multiple notes:
1. The return value is apparently Boolean.
2. Comparisons of expressions are costly. Instead, I'd suggest that domain_simplify has a bool return type to report whether simplification was performed (my ealier suggestions of using `simplify` instead of `simplify_expr` will help in this regard.
3. Do not replace `i_it->guard` when it hasn't even changed.